aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/EndpointBuiltInsDecorator.kt

```diff
@@ -0,0 +1,179 @@
+import software.amazon.smithy.rust.codegen.client.smithy.customize.ClientCodegenDecorator
+import software.amazon.smithy.rust.codegen.client.smithy.endpoint.EndpointCustomization
+import software.amazon.smithy.rust.codegen.client.smithy.endpoint.EndpointRulesetIndex
+import software.amazon.smithy.rust.codegen.client.smithy.endpoint.rustName
+import software.amazon.smithy.rust.codegen.client.smithy.generators.config.ConfigCustomization
+import software.amazon.smithy.rust.codegen.client.smithy.generators.config.ConfigParam
+import software.amazon.smithy.rust.codegen.client.smithy.generators.config.standardConfigParam
+import software.amazon.smithy.rust.codegen.core.rustlang.Writable
+import software.amazon.smithy.rust.codegen.core.rustlang.docs
+import software.amazon.smithy.rust.codegen.core.rustlang.rust
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.writable
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.smithy.customize.AdHocSection
+import software.amazon.smithy.rust.codegen.core.smithy.customize.Section
+import software.amazon.smithy.rust.codegen.core.util.PANIC
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.extendIf
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import java.util.Optional
+
+/** load a builtIn parameter from a ruleset by name */
+fun EndpointRuleSet.getBuiltIn(builtIn: String) = parameters.toList().find { it.builtIn == Optional.of(builtIn) }
+
+/** load a builtIn parameter from a ruleset. The returned builtIn is the one defined in the ruleset (including latest docs, etc.) */
+fun EndpointRuleSet.getBuiltIn(builtIn: Parameter) = getBuiltIn(builtIn.builtIn.orNull()!!)
+fun ClientCodegenContext.getBuiltIn(builtIn: Parameter): Parameter? = getBuiltIn(builtIn.builtIn.orNull()!!)
+fun ClientCodegenContext.getBuiltIn(builtIn: String): Parameter? {
+    val idx = EndpointRulesetIndex.of(model)
+    val rules = idx.endpointRulesForService(serviceShape) ?: return null
+    return rules.getBuiltIn(builtIn)
+}
+
+private fun toConfigParam(parameter: Parameter): ConfigParam = ConfigParam(
+    parameter.name.rustName(),
+    when (parameter.type!!) {
+        ParameterType.STRING -> RuntimeType.String.toSymbol()
+        ParameterType.BOOLEAN -> RuntimeType.Bool.toSymbol()
+    },
+    parameter.documentation.orNull()?.let { writable { docs(it) } },
+)
+
+fun Model.loadBuiltIn(serviceId: ShapeId, builtInSrc: Parameter): Parameter? {
+    val model = this
+    val idx = EndpointRulesetIndex.of(model)
+    val service = model.expectShape(serviceId, ServiceShape::class.java)
+    val rules = idx.endpointRulesForService(service) ?: return null
+    // load the builtIn with a matching name from the ruleset allowing for any docs updates
+    return rules.getBuiltIn(builtInSrc)
+}
```

Comment:
Since `fun EndpointRuleSet.getBuiltIn(builtIn: Parameter) = getBuiltIn(builtIn.builtIn.orNull()!!)` has the non-null assertion, then the return type here can be `Parameter` too.

Reply:
I don't think so? this can return null because a model may not have a built in